### ISSUE TITLE
LSTM MODEL

### ISSUE THREAD

**Issue by akmalramdani** [akmalramdani/AI---Process-Safety#1]
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
from sklearn.metrics import precision_score, f1_score
import matplotlib.pyplot as plt

N_train = 100
N_val = 20
T = 50
F = 3

np.random.seed(42)
X_train = np.random.randn(N_train, T, F)
X_val = np.random.randn(N_val, T, F)

y_train = (np.random.rand(N_train, T) > 0.7).astype(float)
y_val = (np.random.rand(N_val, T) > 0.7).astype(float)

class TimeSeriesDataset(Dataset):
    def __init__(self, X, y):
        self.X = torch.tensor(X, dtype=torch.float32)
        self.y = torch.tensor(y, dtype=torch.float32)

    def __len__(self):
        return self.X.shape[0]

    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]

train_dataset = TimeSeriesDataset(X_train, y_train)
val_dataset = TimeSeriesDataset(X_val, y_val)

train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=16, shuffle=False)

class LSTMSequenceClassifier(nn.Module):
    def __init__(self, input_size, hidden_size=64, num_layers=2, dropout=0.0, bidirectional=False):
        super().__init__()
        self.num_directions = 2 if bidirectional else 1
        self.lstm = nn.LSTM(
            input_size=input_size,
            hidden_size=hidden_size,
            num_layers=num_layers,
            batch_first=True,
            dropout=dropout if num_layers > 1 else 0.0,
            bidirectional=bidirectional
        )
        self.fc = nn.Linear(hidden_size * self.num_directions, 1)

    def forward(self, x):
        output_all, _ = self.lstm(x)
        logits = self.fc(output_all)
        logits = logits.squeeze(-1)
        return logits

input_size = F
hidden_size = 64
num_layers = 2
dropout = 0.1

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Device:", device)

model = LSTMSequenceClassifier(
    input_size=input_size,
    hidden_size=hidden_size,
    num_layers=num_layers,
    dropout=dropout,
    bidirectional=False
).to(device)

criterion = nn.BCEWithLogitsLoss()
optimizer = optim.Adam(model.parameters(), lr=1e-3)

def train_one_epoch(model, loader, optimizer, criterion, device):
    model.train()
    total_loss = 0.0
    for X_batch, y_batch in loader:
        X_batch = X_batch.to(device)
        y_batch = y_batch.to(device)
        optimizer.zero_grad()
        logits = model(X_batch)
        loss = criterion(logits, y_batch)
        loss.backward()
        optimizer.step()
        total_loss += loss.item() * X_batch.size(0)
    return total_loss / len(loader.dataset)

def evaluate_loss(model, loader, criterion, device):
    model.eval()
    total_loss = 0.0
    with torch.no_grad():
        for X_batch, y_batch in loader:
            X_batch = X_batch.to(device)
            y_batch = y_batch.to(device)
            logits = model(X_batch)
            loss = criterion(logits, y_batch)
            total_loss += loss.item() * X_batch.size(0)
    return total_loss / len(loader.dataset)

n_epochs = 10

for epoch in range(1, n_epochs + 1):
    train_loss = train_one_epoch(model, train_loader, optimizer, criterion, device)
    val_loss = evaluate_loss(model, val_loader, criterion, device)
    print(f"Epoch {epoch:02d} | Train Loss: {train_loss:.4f} | Val Loss: {val_loss:.4f}")

def evaluate_metrics(model, loader, device, threshold=0.5):
    model.eval()
    all_logits = []
    all_targets = []
    with torch.no_grad():
        for X_batch, y_batch in loader:
            X_batch = X_batch.to(device)
            y_batch = y_batch.to(device)
            logits = model(X_batch)
            all_logits.append(logits.cpu())
            all_targets.append(y_batch.cpu())
    logits = torch.cat(all_logits, dim=0).numpy().ravel()
    targets = torch.cat(all_targets, dim=0).numpy().ravel()
    probs = 1.0 / (1.0 + np.exp(-logits))
    preds = (probs >= threshold).astype(int)
    precision = precision_score(targets, preds, zero_division=0)
    f1 = f1_score(targets, preds, zero_division=0)
    return precision, f1, probs, targets, preds

precision, f1, probs_val, targets_val, preds_val = evaluate_metrics(
    model, val_loader, device, threshold=0.5
)

print("===================================================")
print(f"Precision (val): {precision:.4f}")
print(f"F1 Score  (val): {f1:.4f}")
print("===================================================")

def plot_time_series_example(model, dataset, idx, device, threshold=0.5):
    model.eval()
    X_seq, y_seq = dataset[idx]
    X_seq_batch = X_seq.unsqueeze(0).to(device)
    with torch.no_grad():
        logits = model(X_seq_batch)
        logits = logits.squeeze(0).cpu().numpy()
    probs = 1.0 / (1.0 + np.exp(-logits))
    preds = (probs >= threshold).astype(int)
    T_len = len(y_seq)
    t_axis = np.arange(T_len)
    plt.figure(figsize=(10, 5))
    plt.plot(t_axis, y_seq.numpy(), marker='o', linestyle='-', label='True label (0/1)')
    plt.plot(t_axis, probs, marker='x', linestyle='--', label='Predicted prob (class=1)')
    plt.step(t_axis, preds, where='mid', alpha=0.5, label='Predicted class (0/1)')
    plt.ylim(-0.1, 1.1)
    plt.xlabel('Time step')
    plt.ylabel('Label / Probability')
    plt.title('Time-series classification: True vs Predicted')
    plt.grid(True)
    plt.legend()
    plt.tight_layout()
    plt.show()

plot_time_series_example(model, val_dataset, idx=0, device=device)
